In [1]:
%%javascript
/**********************************************************************************************
Known Mathjax Issue with Chrome - a rounding issue adds a border to the right of mathjax markup
https://github.com/mathjax/MathJax/issues/1300
A quick hack to fix this based on stackoverflow discussions: 
http://stackoverflow.com/questions/34277967/chrome-rendering-mathjax-equations-with-a-trailing-vertical-line
**********************************************************************************************/

$('.math>span').css("border-left-color","transparent")

<IPython.core.display.Javascript object>

In [2]:
# Find and load Spark context

import os
import sys

import pyspark
from pyspark.sql import SQLContext


app_name = "example-logs"
master = "local[*]"
conf = pyspark.SparkConf().setAppName(app_name).setMaster(master)
sc = pyspark.SparkContext(conf=conf)
sqlContext = SQLContext(sc)

print(sc)
print(sqlContext)

import dateutil.parser
import dateutil.relativedelta as dateutil_rd

In [3]:
%reload_ext autoreload
%autoreload 2

# DAMLAS - Machine Learning At Scale
## Assignment - HW4
Data Analytics and Machine Learning at Scale
Target, Minneapolis

---
__Name:__  Scott Zuehlke   
__Class:__ DAMLAS Summer 2016    
__Email:__ Scott.Zuehlke@Target.com     
__Week:__   04

# Table of Contents <a name="TOC"></a> 

1.  [HW Introduction](#1)   
2.  [HW References](#2)
3.  [HW 4 Problems](#3)   
    4.0.  [Final Project description](#4.0)   
    4.1.  [Build a decision to predict whether you can play tennis or no](#4.1)   
    4.2.  [Regression Tree (OPTIONAL Homework)](#4.2)    
    4.3.  [Predict survival on the Titanic](#4.3)    
    4.4.  [Heritage Healthcare Prize (Predict # Days in Hospital next year)](#4.4)  


<a name="1">
# 1 Instructions
[Back to Table of Contents](#TOC)
* Homework submissions are due by Thursday, 08/18/2016 at 11AM (CT).


* Prepare a single Jupyter notebook (not a requirment), please include questions, and question numbers in the questions and in the responses.
Submit your homework notebook via the following form:

   + [Submission Link - Google Form](http://goo.gl/forms/er3OFr5eCMWDngB72)


### Documents:
* IPython Notebook, published and viewable online.
* PDF export of IPython Notebook.

<a name="2">
# 2 Useful References
[Back to Table of Contents](#TOC)

* [Lecture Slides on Decision Trees and Ensembles](https://www.dropbox.com/s/lm4vuocqoq6mq7k/Lecture-13-Decision-Trees-PLanet.pdf?dl=0)

* Chapter 17 on decision Trees,   https://www.dropbox.com/s/5ca98ah5chqlcmn/Data_Science_from_Scratch%20%281%29.pdf?dl=0   [Please do not share this PDF]
* Karau, Holden, Konwinski, Andy, Wendell, Patrick, & Zaharia, Matei. (2015). Learning Spark: Lightning-fast big data analysis. Sebastopol, CA: O’Reilly Publishers.
* Hastie, Trevor, Tibshirani, Robert, & Friedman, Jerome. (2009). The elements of statistical learning: Data mining, inference, and prediction (2nd ed.). Stanford, CA: Springer Science+Business Media. __(Download for free [here](http://statweb.stanford.edu/~tibs/ElemStatLearn/printings/ESLII_print10.pdf))__
* Ryza, Sandy, Laserson, Uri, Owen, Sean, & Wills, Josh. (2015). Advanced analytics with Spark: Patterns for learning from data at scale. Sebastopol, CA: O’Reilly Publishers.
---

---

## 3.  HW4  <a name="4"></a>
[Back to Table of Contents](#TOC)

 <a name="4.0"></a>
## HW4.0 Final Project description

Please prepare your project description using the following format
* 200 words abstract
* data source and description
* pipeline of steps (in a block diagram)
* Metrics for success

PLEASE NOTE: We will probably have project team sizes of 3 people plus/minus 1

 <a name="4.1"></a>
## HW4.1 Build a decision to predict whether you can play tennis or not

[Back to Table of Contents](#TOC)

Decision Trees

Write a program in Python (or in Spark; this part is optional) to implement the ID3 decision tree algorithm. You should build a tree to predict PlayTennis, based on the other attributes (but, do not use the Day attribute in your tree.). You should read in a space delimited dataset in a file called dataset.txt and output to the screen your decision tree and the training set accuracy in some readable format. For example, here is the tennis dataset. The first line will contain the names of the fields:

<PRE>
Day outlook temperature humidity wind playtennis
d1 sunny hot high FALSE no
d2 sunny hot high TRUE no
d3 overcast hot high FALSE yes
d4 rainy mild high FALSE yes
d5 rainy cool normal FALSE yes
d6 rainy cool normal TRUE no
d6 overcast cool normal TRUE yes
d7 sunny mild high FALSE no
d8 sunny cool normal FALSE yes
d9 rainy mild normal FALSE yes
d10 sunny mild normal TRUE yes
d11 overcast mild high TRUE yes
d12 overcast hot normal FALSE yes
d12 rainy mild high TRUE no
</PRE>

The last column is the classification attribute, and will always contain contain the values yes or no.

For output, you can choose how to draw the tree so long as it is clear what the tree is. You might find it easier if you turn the decision tree on its side, and use indentation to show levels of the tree as it grows from the left. For example:

<PRE>
outlook = sunny
|  humidity = high: no
|  humidity = normal: yes
outlook = overcast: yes
outlook = rainy
|  windy = TRUE: no
|  windy = FALSE: yes

</PRE>

You don't need to make your tree output look exactly like above: feel free to print out something similarly readable if you think it is easier to code.

You may find Python dictionaries especially useful here, as they will give you a quick an easy way to help manage counting the number of times you see a particular attribute.

Here are some FAQs that I've gotten in the past regarding this assignment, and some I might get if I don't answer them now.

__Should my code work for other datasets besides the tennis dataset?__ 
Yes. We will give your program a different dataset to try it out with. You may assume that our dataset is correct and well-formatted, but you should not make assumptions regrading number of rows, number of columns, or values that will appear within. The last column will also be the classification, and will always contain yes or no values.

__Is it possible that some value, like "normal," could appear in more than one column?__
Yes. In addition to the column "humidity", we might have had another column called "skycolor" which could have values "normal," "weird," and "bizarre."

__Could "yes" and "no" appear as possible values in columns other than the classification column?__
Yes. In addition to the classification column "playtennis," we might have had another column called "seasonalweather" which would contain "yes" and "no."

### Write data set to text file.

In [4]:
%%writefile playtennis41.txt
Day outlook temperature humidity wind playtennis
d1 sunny hot high FALSE no
d2 sunny hot high TRUE no
d3 overcast hot high FALSE yes
d4 rainy mild high FALSE yes
d5 rainy cool normal FALSE yes
d6 rainy cool normal TRUE no
d6 overcast cool normal TRUE yes
d7 sunny mild high FALSE no
d8 sunny cool normal FALSE yes
d9 rainy mild normal FALSE yes
d10 sunny mild normal TRUE yes
d11 overcast mild high TRUE yes
d12 overcast hot normal FALSE yes
d13 rainy mild high TRUE no

Overwriting playtennis41.txt


#### View data as data frame

In [5]:
import pandas
pandas.read_table('playtennis41.txt', sep = " ")

Day   outlook temperature humidity   wind playtennis
0    d1     sunny         hot     high  False         no
1    d2     sunny         hot     high   True         no
2    d3  overcast         hot     high  False        yes
3    d4     rainy        mild     high  False        yes
4    d5     rainy        cool   normal  False        yes
5    d6     rainy        cool   normal   True         no
6    d6  overcast        cool   normal   True        yes
7    d7     sunny        mild     high  False         no
8    d8     sunny        cool   normal  False        yes
9    d9     rainy        mild   normal  False        yes
10  d10     sunny        mild   normal   True        yes
11  d11  overcast        mild     high   True        yes
12  d12  overcast         hot   normal  False        yes
13  d13     rainy        mild     high   True         no

### Build Decision Tree by components

In [18]:
from __future__ import division
from collections import Counter, defaultdict
from functools import partial
import math, random
import pandas

def entropy(class_probabilities):
    """given a list of class probabilities, compute the entropy"""
    return sum(-p * math.log(p, 2) for p in class_probabilities if p)

def class_probabilities(labels):
    total_count = len(labels)
    return [count/total_count for count in Counter(labels).values()]

def data_entropy(labeled_data):        
    labels = [label for _, label in labeled_data]
    probabilities = class_probabilities(labels)
    return entropy(probabilities)

def partition_entropy(subsets):
    """find the entropy from this partition of data into subsets"""
    total_count = sum(len(subset) for subset in subsets)    
    return sum(data_entropy(subset)*len(subset)/total_count for subset in subsets )

def inputformat(fileName, delim, classifycolumn, differentiator):
    input_df = pandas.read_table(fileName, sep=delim)
    del input_df[differentiator]
    inputs = input_df.iloc[:,:len(input_df.columns)-1].T.to_dict().values()
    labels = input_df[[len(input_df.columns)-1]].T.to_dict().values()
    formattedInput = list(zip(inputs, [d[classifycolumn] for d in labels]))
    return formattedInput

def group_by(items, key_fn):
    """returns a defaultdict(list), where each input item 
    is in the list whose key is key_fn(item)"""
    groups = defaultdict(list)
    for item in items:
        key = key_fn(item)
        groups[key].append(item)
    return groups
    
def partition_by(inputs, attribute):
    """returns a dict of inputs partitioned by the attribute
    each input is a pair (attribute_dict, label)"""
    return group_by(inputs, lambda x: x[0][attribute]) 

def partition_entropy_by(inputs, attribute):
    """computes the entropy corresponding to the given partition"""        
    partitions = partition_by(inputs, attribute)
    return partition_entropy(partitions.values())


In [28]:
inputs = inputformat('playtennis41.txt', delim=" ", classifycolumn = 'playtennis', differentiator = 'Day')

In [29]:
def build_tree_id3(inputs, split_candidates=None):

    # if this is our first pass, 
    # all keys of the first input are split candidates
    if split_candidates is None:
        split_candidates = inputs[0][0].keys()

    # count Trues and Falses in the inputs
    num_inputs = len(inputs)
    num_trues = [label for item, label in inputs if label].count('yes')
    num_falses = num_inputs - num_trues
    
    if num_trues == 0:                  # if only Falses are left
        return False                    # return a "False" leaf
        
    if num_falses == 0:                 # if only Trues are left
        return True                     # return a "True" leaf

    if not split_candidates:            # if no split candidates left
        return num_trues >= num_falses  # return the majority leaf
                            
    # otherwise, split on the best attribute
    best_attribute = min(split_candidates,
        key=partial(partition_entropy_by, inputs))

    partitions = partition_by(inputs, best_attribute)
    new_candidates = [a for a in split_candidates 
                      if a != best_attribute]
    
    # recursively build the subtrees
    subtrees = { attribute : build_tree_id3(subset, new_candidates)
                 for attribute, subset in partitions.iteritems() }

    subtrees[None] = num_trues > num_falses # default case
    
    return (best_attribute, subtrees)



tree = build_tree_id3(inputs)
print(tree)

('outlook', {'rainy': ('humidity', {'high': ('wind', {False: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), True: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), None: True}), None: True, 'normal': ('wind', {False: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), True: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), None: True})}), 'overcast': ('humidity', {'high': ('wind', {False: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), True: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), None: True}), None: True, 'normal': ('wind', {False: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), True: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), None: True})}), 'sunny': ('humidity', {'high': ('wind', {False: ('temperature', {None: True, 'hot': True, 'mild': True, 'cool': True}), True: ('temperature', {None: T

In [30]:
def classify(tree, input):
    """classify the input using the given decision tree"""
    
    # if this is a leaf node, return its value
    if tree in [True, False]:
        return tree
   
    # otherwise find the correct subtree
    attribute, subtree_dict = tree
    
    subtree_key = input.get(attribute)  # None if input is missing attribute

    if subtree_key not in subtree_dict: # if no subtree for key,
        subtree_key = None              # we'll use the None subtree
    
    subtree = subtree_dict[subtree_key] # choose the appropriate subtree
    return classify(subtree, input)     # and use it to classify the input

### Bulleted View of Decision Tree:
* Split on Outlook

    * If the outlook is overcast, always yes.
    * If the outlook is sunny, we split on humidity.
        * If humidity is High, always no.
        * If humidity is Normal, always yes.
    * If the outlook is rainy, we split on wind.
        * If wind is strong, always no.
        * If wind is weak, always yes.

###  Visual View of Decision Tree

                                     Outlook
                                    /   |   \
                                   /    |    \
                                  /     |     \
                               Sunny Overcast Rainy
                              /         |         \
                             /          |          \
                            /           |           \
                       Humidity         |          Wind
                        /  \            |          /   \
                       /    \           |         /     \
                      /      \          |        /       \
                  High      Normal      |      True    False
                   /             \      |      /           \ 
                  /               \     |     /             \
                 /                 \    |    /               \
                NO                YES  YES  NO               YES

### Pythonic View of Decision Tree

In [9]:
('outlook', 
 {'rainy': ('wind', {False: True, True: False, None: True}), 
  'overcast': True, 
  'sunny': ('humidity', {'high': False, None: False, 'normal': True}), None: True})

('outlook',
 {None: True,
  'overcast': True,
  'rainy': ('wind', {None: True, False: True, True: False}),
  'sunny': ('humidity', {None: False, 'high': False, 'normal': True})})

#### Now to build the tree for classification.

In [31]:
for i in inputs:
    
    print(classify(tree, i[0]))

True
True
True
True
True
True
True
True
True
True
True
True
True
True


__ HW4.1.1 What is the classification accuracy of the tree on the training data?__


The tree predicts everything with perfect accuracy.  This makes sense given that we're building on the entire data set (with every possibility) and everything ends in either a true or a false.

__HW4.1.2  Is it possible to produce some set of correct training examples that will get the algorihtm
to include the attribute Temperature in the learned tree, even though the true target concept is
independent of Temperature? if no, explain. If yes, give such a set. __

__HW4.1.3  Now, build a tree using only examples D1–D7. What is the classification accuracy for the
training set? what is the accuracy for the test set (examples D8–D14)? explain why you think these
are the results.__

In [13]:
%%writefile playtennis_413train.txt
Day outlook temperature humidity wind playtennis
d1 sunny hot high FALSE no
d2 sunny hot high TRUE no
d3 overcast hot high FALSE yes
d4 rainy mild high FALSE yes
d5 rainy cool normal FALSE yes
d6 rainy cool normal TRUE no
d6 overcast cool normal TRUE yes
d7 sunny mild high FALSE no

Overwriting playtennis_413train.txt


In [14]:
import pandas
pandas.read_table('playtennis_413train.txt', sep = " ")

Day   outlook temperature humidity   wind playtennis
0  d1     sunny         hot     high  False         no
1  d2     sunny         hot     high   True         no
2  d3  overcast         hot     high  False        yes
3  d4     rainy        mild     high  False        yes
4  d5     rainy        cool   normal  False        yes
5  d6     rainy        cool   normal   True         no
6  d6  overcast        cool   normal   True        yes
7  d7     sunny        mild     high  False         no

In [15]:
inputs = inputformat('playtennis_413train.txt', delim=" ", classifycolumn = 'playtennis', differentiator = 'Day')
step1lowestentropy = lowest_entropy(['outlook', 'temperature', 'humidity', 'wind'])
print(step1lowestentropy)

('outlook', 0.3443609377704336)


In [16]:
sunny_inputs = [(input, label) for input, label in inputs if input["outlook"] == "sunny"]
overcast_inputs = [(input, label) for input, label in inputs if input["outlook"] == "overcast"]
rain_inputs = [(input, label) for input, label in inputs if input["outlook"] == "rainy"]

for key in ['temperature', 'humidity', 'wind']:
    print 'Sunny', key, partition_entropy_by(sunny_inputs, key)
    
print(' '*80)

for key in ['temperature', 'humidity', 'wind']:
    print 'Overcast', key, partition_entropy_by(overcast_inputs, key)
    
print(' '*80)

for key in ['temperature', 'humidity', 'wind']:
    print 'Rainy', key, partition_entropy_by(rain_inputs, key)

Sunny temperature 0.0
Sunny humidity 0.0
Sunny wind 0.0
                                                                                
Overcast temperature 0.0
Overcast humidity 0.0
Overcast wind 0.0
                                                                                
Rainy temperature 0.666666666667
Rainy humidity 0.666666666667
Rainy wind 0.0


### Bulleted View of Decision Tree:
* Split on Outlook

    * If the outlook is overcast, always yes.
    * If the outlook is sunny, always no.
    * If the outlook is rainy, we split on wind.
        * If wind is strong, always no.
        * If wind is weak, always yes.

###  Visual View of Decision Tree

                                     Outlook
                                    /   |   \
                                   /    |    \
                                  /     |     \
                               Sunny Overcast Rainy
                               /        |         \
                              /         |          \
                             /          |           \
                            /           |           Wind
                           /            |           /  \ 
                          /             |          /    \
                         /              |         /      \
                        /               |       False    True  
                       /                |       /          \
                      /                 |      /            \
                     /                  |     /              \
                    NO                 YES   YES             NO

In [17]:
tree = build_tree_id3(inputs)
print(tree)

('outlook', {'rainy': ('wind', {False: True, True: False, None: True}), 'overcast': True, 'sunny': False, None: False})


#### Pythonic View of Decision Tree

In [18]:
('outlook', 
 {'rainy': ('wind', {False: True, True: False, None: True}), 
  'overcast': True, 
  'sunny': False, None: False})

('outlook',
 {None: False,
  'overcast': True,
  'rainy': ('wind', {None: True, False: True, True: False}),
  'sunny': False})

In [19]:
print "sunny cool normal FALSE yes: ", classify(tree, 
    {"outlook" : "sunny",
     "temperature" : "cool",
     "humidity" : "normal",
     "wind" : False})

print "rainy mild normal FALSE yes: ", classify(tree, 
    {"outlook" : "rainy",
     "temperature" : "mild",
     "humidity" : "normal",
     "wind" : False})

print "sunny mild normal TRUE yes: ", classify(tree, 
    {"outlook" : "sunny",
     "temperature" : "mild",
     "humidity" : "normal",
     "wind" : True})

print "overcast mild high TRUE yes: ", classify(tree, 
    {"outlook" : "overcast",
     "temperature" : "mild",
     "humidity" : "high",
     "wind" : True})

print "overcast hot normal FALSE yes: ", classify(tree, 
    {"outlook" : "overcast",
     "temperature" : "hot",
     "humidity" : "normal",
     "wind" : False})

print "rainy mild high TRUE no: ", classify(tree, 
    {"outlook" : "rainy",
     "temperature" : "mild",
     "humidity" : "high",
     "wind" : True})

sunny cool normal FALSE yes:  False
rainy mild normal FALSE yes:  True
sunny mild normal TRUE yes:  False
overcast mild high TRUE yes:  True
overcast hot normal FALSE yes:  True
rainy mild high TRUE no:  False


Wrong Right Wrong Right Right Right = 4/6 = 67% accuracy

__HW4.1.4 In this case, and others, there are only a few labelled examples available for training (that
is, no additional data is available for testing or validation). Suggest a concrete pruning strategy, that
can be readily embedded in the algorithm, to avoid over fitting. Explain why you think this strategy
should work.__

 <a name="4.3"></a>
## HW4.3 Predict survival on the Titanic using Python (Logistic regression, SVMs, Random Forests)

[Back to Table of Contents](#TOC)

The sinking of the RMS Titanic is one of the most infamous shipwrecks in history.  On April 15, 1912, during her maiden voyage, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. This sensational tragedy shocked the international community and led to better safety regulations for ships.

One of the reasons that the shipwreck led to such loss of life was that there were not enough lifeboats for the passengers and crew. Although there was some element of luck involved in surviving the sinking, some groups of people were more likely to survive than others, such as women, children, and the upper-class.

In this challenge, you need to review (and edit the code) in this [notebook](http://nbviewer.jupyter.org/urls/dl.dropbox.com/s/kmbgrkhh73931lo/Titanic-EDA-LogisticRegression.ipynb) to do analysis of what sorts of people were likely to survive. In particular, please look at how the tools of machine learning are used to predict which passengers survived the tragedy. Please share any usefule graphs/analysis you come up with via the group email.

For more details see:

* https://www.kaggle.com/c/titanic

 <a name="4.4"></a>
 ## HW4.4 Heritage Healthcare Prize (Predict # Days in Hospital next year)
[Back to Table of Contents](#TOC)

1. Introduction 
Back to Table of Contents

The Heritage Health Prize (HHP) was a data science challenge sponsored by The Heritage Provider Network. It took place from April 4, 2011 to April 4, 2013. For information on the winning entries, please see here.

Please see the following notebooks for more background and candidate solutions


- Spark Map-Reduce + MMLlib solution (with optional extensions) See [Notebook](http://nbviewer.jupyter.org/urls/dl.dropbox.com/s/v52cxipe7yftf97/HeritageHealthPrizeUnitTestNotebook_Spark-Map-Reduce.ipynb)

- Spark SQL + MLLib solution (with optional extensions): [Notebook](http://nbviewer.jupyter.org/urls/dl.dropbox.com/s/s2wxg6g982oho5m/HeritageHealthPrizeUnitTestNotebook_SQL_FINAL.ipynb)


Please look at section 7 in both notebooks complete any one or more the suggested next steps. E.g.,

* Please complete the EDA extensions using inspiration from the Titanic Notebook from above.
* __Complete Section 3.B: EDA-0. Gather information to see what transformations may need to be done on the data.__
Answer questions about each raw DataFrame. In general, is the data in good shape? For example, in each of the Target DataFrames (df_target_Y1, df_target_Y2, df_target_Y3), what values does DaysInHospital take on? Are they all integers? What values does ClaimsTruncated take on? Are they all integers? In the Claims DataFrame (df_claims), how many different ProviderIDs are there? How many different PrimaryConditionGroups are there? What are their values? What values can the CharlesonIndex take on? Are they integers? In the Drug Count DataFrame (df_drug_count), what values can DrugCount take on? Are they all integers? Given this information, what transformations are needed?

* __Complete Section 3.D: EDA-1. Create tables and graphs to display information about the transformed DataFrames. __
For inspiration, see the Titanic notebook discussed above. Answer questions about each DataFrame. For example, in each of the Target DataFrames (df_target_Y1, df_target_Y2, df_target_Y3), what is the minimum, maximum, mean, and standard deviation of DaysInHospital? In the Claims DataFrame, group by MemberID and Year and count the number of records. What is the minimum, maximum, mean, and standard deviation of the count? Do the same for the Drug Count and Lab Count DataFrames, etc.


* __ Please generate ensemble of DT model using 100 trees with 8 nodes and report the Loss __
Try additional models. See possibilities here (e.g. Decision Tree Regressor, Gradient-Boosted Trees Regressor, Random Forest Regressor). See an example here. Tune their hyperparameters. Try different feature selections. Try a two-step model.
